### ISSUE TITLE
Urgent: I am getting this error

### ISSUE THREAD

**Issue by aviagga2** [Baroshem/nuxt-security#309]
Refused to execute inline script because it violates the following Content Security Policy directive: "script-src 'nonce-SXSBo/Czk+z63fJcSTTalg==' 'strict-dynamic' 'nonce-d01b6c42136ec8749b0b830919f78893' https://me.kis.v2.scr.kaspersky-labs.com wss://me.kis.v2.scr.kaspersky-labs.com". Either the 'unsafe-inline' keyword, a hash ('sha256-XWbkakumASHI71QQKyQ47iCxsPxcNJtxovXPSycsqcY='), or a nonce ('nonce-...') is required to enable inline execution.

**Comment by Baroshem** [comment]
Hey,

It is not a bug, it is an intended behavior.

Check out the documentation about the usage of CSP and how to adjust your app to make it work with Strict and Secure CSP https://nuxt-security.vercel.app/documentation/advanced/strict-csp#allowing-inline-elements-for-nuxt

**Comment by aviagga2** [comment]
Can you please guide me on how to resolve this? I get this issue when I proxy my website using Cloudflare. It's a request, please.

**Comment by Baroshem** [comment]
How do you execute this script? You load it globally using nuxt config, useHead, or in the template?

**Comment by aviagga2** [comment]
security: {
    headers: {
      crossOriginEmbedderPolicy: 'unsafe-none',
      contentSecurityPolicy: {
        'img-src': ["'self'", 'data:']
      },
    },
  },
  
  
  this is my nuxt.config

**Comment by aviagga2** [comment]
@vejja can you please guide me...

**Comment by vejja** [comment]
Hi @aviagga2 can you let me know 
- which script is being denied, in the issues tab of Chrome
- The value of your CSP header in the network tab when you load your page
- Who is adding the second nonce (if you know)

**Comment by aviagga2** [comment]
I also want to enable Cloudflare Zaraz for Analytics, but if I enable that, I get many more errors. Can you please guide me how to resolve them.

Page URL: https://www.a23d.co/

![image](https://github.com/Baroshem/nuxt-security/assets/152681873/65bdc481-d33c-4096-acea-e558a34c3e32)

**Comment by vejja** [comment]
@aviagga2 If you disable Zaraz do you still have issues ?

**Comment by aviagga2** [comment]
Yes

**Comment by aviagga2** [comment]
So should I add script-src self here:

security: {
headers: {
crossOriginEmbedderPolicy: 'unsafe-none',
contentSecurityPolicy: {
'img-src': ["'self'", 'data:']
},
},
},

**Comment by vejja** [comment]
> So should I add script-src self here:
> 
> security: { headers: { crossOriginEmbedderPolicy: 'unsafe-none', contentSecurityPolicy: { 'img-src': ["'self'", 'data:'] }, }, },

Not right now, I want to know if we can make it work without changes
Cloudflare claim that they are able to parse the nonce and inject it in their inline script, so it should work if I read their doc correctly
Note: if you are a free customer for Cloudflare, Javascript Detection is always enabled

**Comment by aviagga2** [comment]
No I have a Pro account. And if I try to disable it I get a warning.

![image](https://github.com/Baroshem/nuxt-security/assets/152681873/49836fe4-3761-438a-8fde-a213ee68e603)

**Comment by vejja** [comment]
There is something that I do not understand with Cloudflare

1. I can also that Nuxt Security inserts the nonce in the CSP header
![Capture d’écran 2023-12-02 à 14 26 37](https://github.com/Baroshem/nuxt-security/assets/7295259/0239aa0b-eeea-4c1e-877e-4a3ea6d02e02)

2. I can see that Nuxt Security insert the nonces everywhere except from the Cloudflare script (this is normal because Cloudflare injection happens after the page is generated by Nuxt)
![Capture d’écran 2023-12-02 à 14 28 49](https://github.com/Baroshem/nuxt-security/assets/7295259/439a52de-3091-4f39-a56a-d938d37e3e90)

3. I also read that Cloudflare will detect the nonce and inject it themselves
![Capture d’écran 2023-12-02 à 14 29 44](https://github.com/Baroshem/nuxt-security/assets/7295259/210a8e61-5ea2-4662-8abd-491a6f27bc26)

4. But I can see that they inject their script at the bottom of your page without any nonce
![Capture d’écran 2023-12-02 à 14 24 01](https://github.com/Baroshem/nuxt-security/assets/7295259/aa1aaef2-024d-4a4e-9b74-763fc1faec95)

So 3. and 4. do not seem to work as intended

**Comment by aviagga2** [comment]
You are absolutely right, it should work that way if Cloudflare inserts script with nonce.

The issue we see in the console marks the script at line 18:
![image](https://github.com/Baroshem/nuxt-security/assets/152681873/f507c681-5f85-4c7e-a427-c0dbf96b2b2d)

I don't understand what exactly is happening here.

I want to thank you in advance for all the help and your time.

**Comment by vejja** [comment]
The parser indicates line 18 and positions the red cross there (on the Nuxt script) but in fact it is the second script block of line 18 (the Cloudflare JavaScript Detection) which is disabled.
Can you ask Cloudflare why they do not detect the nonce ?

Also, can you try modifying Nuxt Security the following way: 
```
security: {
  headers: {
    crossOriginEmbedderPolicy: 'unsafe-none',
    contentSecurityPolicy: {
      'img-src': ["'self'", 'data:'],
      'script-src': ["'strict-dynamic'", "'nonce-{{nonce}}'"]
    },
  },
},
```
I am wondering whether the fact that we have 'unsafe-inline' in the directives confuses the Cloudflare parser in some way ?

**Comment by aviagga2** [comment]
I did it, but still the same issue.

Do you thing we can/should disable JavaScript Detection?

**Comment by vejja** [comment]
As far as I understand, Javascript Detection is meant to fight bots. Also, I understand that you need to enable Cloudflare Zaraz. So ultimately this is a business decision.

My assessment of this situation is that Nuxt is not going to be compatible with Cloudflare if CSP is enabled. This could change but we would need to spend more time with Cloudflare to understand why they do not detect nonces.

So in summary I would say that you have to very different solutions:

1. Disable CSP entirely, and use Cloudflare functionalities. 
  - If you are using Nuxt Security <= 1.0.0-rc.4, you can do this with `contentSecurityPolicy: ' '` 

2. Enable CSP, but disable Cloudflare functionalities. 
  - You will need to disable Javascript Detection.
  - You will need to also disable Cloudflare Zaraz

**Comment by vejja** [comment]
@aviagga2 actually I believe there is a 3rd solution which might combine the best of both worlds at the expense of a bit more work on your side. 
I read the Cloudflare docs and it looks like it is possible to manually import the Cloudflare scripts and disable the auto-injection of these scripts.

- Zaraz can be loaded manually : https://developers.cloudflare.com/zaraz/advanced/load-zaraz-manually/
- JavaScript Detection might be loadable manually : `The snippets of JavaScript will contain a source pointing to the challenge platform, with paths that start with /cdn-cgi/challenge-platform/...`

So you could disable the auto-injection from the Cloudflare Dashboard, and instead load them via Nuxt's `useHead`:

```ts
useHead({
  scripts: [
    { src: "/cdn-cgi/zaraz/i.js", crossorigin: true, referrerpolicy:"origin" },
    { src: "/cdn-cgi/challenge-platform/scripts/jsd/main.js", crossorigin: true, referrerpolicy: "origin" }
  ]
}, { mode: 'client' })
```

**Comment by vejja** [comment]
> That could be a great solution. Let me try this.
> 
> Can I add it in the layout?
> 
> Actually I tried, but I didn't see any script being added in my head element.

You can use it in any `<script setup>` I believe
Or alternatively, in `nuxt.config.ts` under `app.head`

**Note:** I made a mistake above, it is `script` and not `scripts`

**Comment by aviagga2** [comment]
s.js?z=JTdCJTIyZXhlY3V0ZWQlMjIlM0ElNUIlNUQlMkMlMjJ0JTIyJTNBJTIyQTIzRCUyMC0lMjBUaGUlMjBsYXJnZXN0JTIwM0QlMjBhc3NldCUyMGxpYnJhcnklMjIlMkMlMjJ4JTIyJTNBMC4xNjc3NDA1ODY3MDI2ODkxJTJDJTIydyUyMiUzQTE5MjAlMkMlMjJoJTIyJTNBMTA4MCUyQyUyMmolMjIlM0E5NTglMkMlMjJlJTIyJTNBMTI5NyUyQyUyMmwlMjIlM0ElMjJodHRwcyUzQSUyRiUyRnd3dy5hMjNkLmNvJTJGJTIyJTJDJTIyciUyMiUzQSUyMiUyMiUyQyUyMmslMjIlM0EyNCUyQyUyMm4lMjIlM0ElMjJVVEYtOCUyMiUyQyUyMm8lMjIlM0EtMzMwJTJDJTIycSUyMiUzQSU1QiU1RCU3RA==:1  Error executing script: (function(w,d){w.zarazData.executed.push("Pageview");})(window,document)
 EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self' https: 'unsafe-inline' 'strict-dynamic' 'nonce-CFmpRgaeVE14zbB0/Va7pg==' 'nonce-3da3fd7fba3edb692066e4e23970444e' https://me.kis.v2.scr.kaspersky-labs.com wss://me.kis.v2.scr.kaspersky-labs.com".
 
 Now I get this error

**Comment by vejja** [comment]
Not available right now to better help but can you copy paste our default array values for script-src and just add « ‘unsafe-eval’ » at the end of the array ?

**Comment by aviagga2** [comment]
@vejja Thanks a lot for all your time and help! Really appreciate it...

**Comment by aviagga2** [comment]
@vejja Thanks a lot for all your time and help! Really appreciate it...

**Comment by vejja** [comment]
@aviagga2 did you manage to get Zaraz working with the setup above ?
I'd like to add a FAQ on Zaraz to our doc, and I actually believe it could work without `useHead` after thinking about it over the weekend.

**Comment by Baroshem** [comment]
Hey guys,

Really glad reading the comments here. Huge kudos once again to @vejja. You are doing an amazing work buddy! 🚀

**Comment by avi-agg** [comment]
> @aviagga2 did you manage to get Zaraz working with the setup above ? I'd like to add a FAQ on Zaraz to our doc, and I actually believe it could work without `useHead` after thinking about it over the weekend.

Nope it didn't work. I was getting all those errors again. Finally I decided to not use Zaraz. I am planning to use some other tracking method.

**Comment by vejja** [comment]
> > @aviagga2 did you manage to get Zaraz working with the setup above ? I'd like to add a FAQ on Zaraz to our doc, and I actually believe it could work without `useHead` after thinking about it over the weekend.
> 
> Nope it didn't work. I was getting all those errors again. Finally I decided to not use Zaraz. I am planning to use some other tracking method.

Let me know if you still want to try it. Would love to make it work, and it was great having you iterate so quickly
Cheers

**Comment by Baroshem** [comment]
@avi-agg 

We are really glad to see you that you would like to help! Do not hesitate to open issues and recommend things that could be done better. And also let us know if you would like to contribute in any other way. We will be more than happy to assist you :)
